include/alp/imf.hpp

```diff
@@ -50,107 +50,202 @@
+						s == static_cast< const Strided & >( other ).s;
+				}
+		};
+
+		/**
+		 * The identity IMF.
+		 * \f$I_n = [0, n)\f$
+		 * \f$Id = I_n \rightarrow I_n; i \mapsto i\f$
+		 */
+
+		class Id: public Strided {
+
+			public:
 
-                size_t map(size_t i) {
-                    return f->map( g->map( i ) );
-                }
+				Id( const size_t n ) : Strided( n, n, 0, 1 ) {}
+		};
 
-                Composed(std::shared_ptr<IMF> f, std::shared_ptr<IMF> g): IMF( g->n, f->N ), f(f), g(g) { }
+		class Select: public IMF {
+
+			public:
 
-        };
+				std::vector< size_t > select;
+
+				size_t map( size_t i ) const {
+					std::cout << "Calling Select map; vec size = " << select.size() << std::endl;
+					return select.at( i );
+				}
+
+				Select(size_t N, std::vector< size_t > &select): IMF( select.size(), N ), select( select ) {
+					//if ( *std::max_element( select.cbegin(), select.cend() ) >= N) {
+					//	throw std::runtime_error("IMF Select beyond range.");
+					//}
+				}
+
+				Select(size_t N, std::vector< size_t > &&select): IMF( select.size(), N ), select( select ) {
+					std::cout << "Select move constructor" << std::endl;
+					std::cout << (this->select).size() << std::endl;
+					//if ( *std::max_element( select.cbegin(), select.cend() ) >= N) {
+					//	throw std::runtime_error("IMF Select beyond range.");
+					//}
+				}
+
+				template< typename OtherIMF >
+				bool isSame( const OtherIMF &other ) const {
+					return IMF::isSame( other ) && select == static_cast< const Select & >( other ).select;
+				}
+		};
+
+		/**
+		 * A composition of two IMFs.
+		 * \f$I_{g,n} =[0, n), I_{g,N} =[0, N)\f$
+		 * \f$I_{f,n} =[0, n), I_{f,N} =[0, N)\f$
+		 * \f$Composed_{f, g} = I_{g,n} \rightarrow I_{f,N}; i \mapsto f( g( i ) )\f$
+		 *
+		 * \tparam FirstIMF  The function that is applied first (i.e., \f$g\f$ )
+		 * \tparam SecondIMF The function that is applied second (i.e., \f$f\f$ )
+		 *
+		 * For specific combinations of the IMF types, there are specializations
+		 * that avoid nested function calls by fusing two functions into one.
+		 */
+
+		template< typename FirstIMF, typename SecondIMF >
+		class Composed: public IMF {
+
+			public:
+				const FirstIMF &g;
+				const SecondIMF &f;
+
+				size_t map( const size_t i ) const {
+						std::cout << "Calling Composed< IMF, IMF>::map()" << std::endl;
+						return f.map( g.map( i ) );
+				}
+
+				Composed( const FirstIMF &g, const SecondIMF &f ):
+					IMF( g.n, f.N ), g( g ), f( f ) {}
+
+		};
+
+		template< typename IMF1, typename IMF2 >
+		struct composed_type {
+			typedef Composed< IMF1, IMF2 > type;
+		};
+
+		template<>
+		struct composed_type< Strided, Strided > {
+			typedef Strided type;
+		};
+
+		template<>
+		struct composed_type< Id, Strided > {
+			typedef Strided type;
+		};
+
+		template<>
+		struct composed_type< Strided, Id > {
+			typedef Strided type;
+		};
+
+		template<>
+		struct composed_type< Id, Id > {
```

Comment:
We have Id to avoid writing Strided( 0, N ) instead of Id( N ). Since we do not use dynamic polymorphism, we have to specialize to all variants. Do you suggest completely removing Id?